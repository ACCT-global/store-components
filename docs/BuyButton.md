ðŸ“¢ Don't fork this project. Use, [contribute](https://github.com/vtex-apps/awesome-io#contributing), or open issues through [Store Discussion](https://github.com/vtex-apps/store-discussion).

# Buy Button

`BuyButton` is a VTEX Component that is resposible to handle events of adding products in the minicart. This Component can be imported and used by any VTEX App.

![image](https://user-images.githubusercontent.com/284515/70233985-69e13700-173e-11ea-91f7-6675a6a0e73b.png)

## Configuration

1. Import the vtex.store-component's app to your theme's dependencies in the manifest.json, for example:

```json
  dependencies: {
    "vtex.store-components": "3.x"
  }
```

2. Add the `buy-button` block to any block bellow `store.product`. For example:

```json
  "store.product": {
    "children": [
      "flex-layout.row#product",
    ]
  },
  "flex-layout.row#product": {
    "children": [
      "buy-button"
    ]
  }
```

| Prop name            | Type      | Description                                                                      | Default value      |
| -------------------- | --------- | -------------------------------------------------------------------------------- | ------------------ |
| `isOneClickBuy`      | `Boolean` | Should redirect to the checkout page or not                                      | `false`              |
| `shouldOpenMinicart` | `Boolean` | Should open the Minicart after clicking the button                               | `false`              |
| `large`              | `Boolean` | Sets button to large style, filling whole width (like a `block`)                 | -                  |
| `available`          | `Boolean` | If component is available or not                                                 | `true`               |
| `showToast`          | `Boolean` | If toast with feedback should be shown after add item request is processed       | -                  |
| `showItemsPrice`     | `Boolean` | If you want to show the total price of items to be added to cart                 | `false`              |
| `customToastURL`     | `String`  | Set the link associated with the Toast created when adding an item to your cart. | `/checkout/#/cart` |
| `shouldAddToCart`    | `Boolean` | If it should add to cart when clicked                                             | true          |
| `showTooltipOnSkuNotSelected` | `Boolean` | If it should show a tooltip when you click the button but there's no SKU selected | `true` |

## Customization

| CSS Handles |
| --- |
| `buyButtonText`       |
| `buttonDataContainer` |
| `buttonItemsPrice`    |
