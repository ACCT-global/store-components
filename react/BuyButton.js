import { Component } from 'react'
import PropTypes from 'prop-types'
import { compose, graphql } from 'react-apollo'

import Spinner from '@vtex/styleguide/lib/Spinner'
import spinnerStyle from '@vtex/styleguide/lib/Spinner/style.css'

import productsQuery from './graphql/productsQuery.graphql'


const options = {
  options: ({
    category = '',
    specificationFilters = '',
    priceRange = '',
    collection = '',
    orderBy = '',
    from = 0,
    to = 8,
    salesChannel = '',
  }) => ({
    variables: {
      category,
      specificationFilters: specificationFilters ? [specificationFilters] : [],
      priceRange,
      collection,
      orderBy,
      from,
      to: to - 1,
      salesChannel,
    },
    ssr: false,
  }),
}


/**
 * BuyButton Component. Adds a list of itens to the cart.
 */
class BuyButton extends Component {

  constructor(props) {
    super(props)
    this.state = { data: { loading: true } }
  }
  
  componentDidMount() {
    this.setState( { data: graphql(productsQuery, options) } )
  }

  render() {
    const { data, maxItems, titleColor, titleText } = this.props
    const products = data['error'] ? [] : data.products
    const { sliderMounted } = this.state
    const slideSettings = this.configureSettings()

    return (
      <div className="ml7 mr7 pv4">
         { loading && (
          <div className="flex justify-around pa7">
            <div className="w3">
              <Spinner style={spinnerStyle} />
            </div>
          </div>
        )}
      </div>
    )
  }
}

Shelf.propTypes = {
  /** The quantity of products to be added to the cart */
  quantity: PropTypes.number.isRequired,
  /** The specification of wich product will be added to the cart */
  skuId:    PropTypes.string.isRequired,
  /** Wich seller is being referenced by the button */
  seller:   PropTypes.string.isRequired,
  /**  */
  salesChannel: PropTypes.string.isRequired,
  /** Should redirect or not */
  redirect: PropTypes.bool
}

export default BuyButton
