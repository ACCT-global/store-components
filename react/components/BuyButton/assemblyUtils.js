import { pathOr } from 'ramda'

export const sumAssembliesPrice = assemblyOptions => {
  const cleanAssemblies = assemblyOptions || {}
  const assembliesGroupItems = Object.values(cleanAssemblies)
  return assembliesGroupItems.reduce((sum, groupItems) => {
    const groupPrice = groupItems.reduce((groupSum, item) => {
      const childrenPrice = item.children
        ? sumAssembliesPrice(item.children)
        : 0
      const itemCost = item.price * item.quantity
      return groupSum + itemCost + childrenPrice * item.quantity
    }, 0)
    return groupPrice + sum
  }, 0)
}

export const transformAssemblyOptions = (
  assemblyOptions,
  parentPrice,
  parentQuantity
) => {
  const cleanAssemblies = assemblyOptions || {}
  const assembliesKeys = Object.keys(cleanAssemblies)
  const options = [] // contains options sent as arguments to graphql mutation
  const added = [] // array with added assemblies data to show in minicart optimistic preview
  const removed = [] // array with removed assemblies data to show in minicart optimistic preview
  for (const groupId of assembliesKeys) {
    const items = cleanAssemblies[groupId]
    for (const item of items) {
      const {
        id,
        quantity,
        seller,
        initialQuantity,
        choiceType,
        name,
        price,
        children,
      } = item
      const childrenAddedData = children
        ? transformAssemblyOptions(
            children,
            item.price,
            item.quantity * parentQuantity
          )
        : null

      const addedChildrenCount = pathOr(
        0,
        ['options', 'length'],
        childrenAddedData
      )

      const {
        options: childrenOptions,
        assemblyOptions: childrenAssemblyOptions,
      } = childrenAddedData || {}

      if (quantity >= initialQuantity && quantity > 0) {
        added.push({
          normalizedQuantity: quantity,
          extraQuantity: quantity - initialQuantity,
          choiceType,
          item: {
            name,
            sellingPrice: price,
            quantity,
            sellingPriceWithAssemblies:
              price + sumAssembliesPrice(children || {}),
            id,
            ...(childrenAssemblyOptions
              ? { assemblyOptions: childrenAssemblyOptions }
              : {}),
          },
        })
      }

      if (quantity < initialQuantity && choiceType === 'TOGGLE') {
        removed.push({
          name,
          initialQuantity,
          removedQuantity: initialQuantity - quantity,
        })
      }

      if (quantity !== initialQuantity || addedChildrenCount > 0) {
        options.push({
          assemblyId: groupId,
          id,
          quantity: quantity * parentQuantity,
          seller,
          ...(childrenOptions && childrenOptions.length > 0
            ? { options: childrenOptions }
            : {}),
        })
      }
    }
  }

  return {
    options,
    assemblyOptions: {
      added,
      removed,
      parentPrice,
    },
  }
}
