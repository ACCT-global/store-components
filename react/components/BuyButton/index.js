import PropTypes from 'prop-types'
import React, { Component, Fragment } from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import { injectIntl, intlShape, FormattedMessage } from 'react-intl'
import ContentLoader from 'react-content-loader'
import { compose, pathOr, pick } from 'ramda'

import { Button, withToast } from 'vtex.styleguide'

const CONSTANTS = {
  SUCCESS_MESSAGE_ID: 'buybutton.buy-success',
  ERROR_MESSAGE_ID: 'buybutton.add-failure',
  CHECKOUT_URL: '/checkout/#/cart',
  TOAST_TIMEOUT: 3000,
}

/**
 * BuyButton Component.
 * Adds a list of sku items to the cart.
 */
export class BuyButton extends Component {
  static defaultProps = {
    isOneClickBuy: false,
    available: true,
  }

  state = {
    isLoading: false,
    isAddingToCart: false,
    timeOut: null,
  }

  translateMessage = id => this.props.intl.formatMessage({ id: id })

  toastMessage = success => {
    const message = success
      ? this.translateMessage(CONSTANTS.SUCCESS_MESSAGE_ID)
      : this.translateMessage(CONSTANTS.ERROR_MESSAGE_ID)
    this.props.showToast({ message })
  }

  skuItemToMinicartItem = ({
    skuId: id,
    variant: skuName,
    price: sellingPrice,
    ...restSkuItem
  }) => {
    return {
      id,
      sellingPrice,
      skuName,
      ...pick(
        [
          'detailUrl',
          'imageUrl',
          'quantity',
          'seller',
          'name',
          'options',
          'listPrice',
        ],
        restSkuItem
      ),
      index: 1,
    }
  }

  handleAddToCart = async () => {
    const {
      addToCart,
      skuItems,
      isOneClickBuy,
      onAddStart,
      onAddFinish,
    } = this.props
    this.setState({ isAddingToCart: true })
    onAddStart && onAddStart()

    let showToastMessage = null
    try {
      const minicartItems = skuItems.map(this.skuItemToMinicartItem)

      const response = await addToCart(minicartItems)
      console.info(response)
      const linkStateItems = pathOr([], ['data', 'addToCart'], response)

      const success = skuItems.map(skuItem =>
        !!linkStateItems.find(({ id }) => id === skuItem.skuId)
      )

      if (isOneClickBuy) location.assign(CONSTANTS.CHECKOUT_URL)
      showToastMessage = () => this.toastMessage(success.length >= 1)
    } catch (err) {
      console.error(err)
      showToastMessage = () => this.toastMessage(false)
    }

    setTimeout(
      () => this.setState({ isAddingToCart: false }, () => {
        showToastMessage()
        onAddFinish && onAddFinish()
      }),
      500
    )
  }

  render() {
    const { children, skuItems, available, large } = this.props
    const loading = this.state.isLoading || !skuItems
    const { isAddingToCart } = this.state

    return (
      <Fragment>
        {loading ? (
          <ContentLoader />
        ) : (
          <Button
            primary
            block={large}
            disabled={!available}
            onClick={this.handleAddToCart}
            isLoading={isAddingToCart}
          >
            {available ? (
              children
            ) : (
              <FormattedMessage id="buyButton-label-unavailable" />
            )}
          </Button>
        )}
      </Fragment>
    )
  }
}

BuyButton.propTypes = {
  /** SKU Items to be added to the cart */
  skuItems: PropTypes.arrayOf(
    PropTypes.shape({
      /** Specification of which product will be added to the cart */
      skuId: PropTypes.string.isRequired,
      /** Quantity of the product sku to be added to the cart */
      quantity: PropTypes.number.isRequired,
      /** Which seller is being referenced by the button */
      seller: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
      name: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      variant: PropTypes.string,
      brand: PropTypes.string.isRequired,
      options: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string.isRequired,
          quantity: PropTypes.number.isRequired,
          assemblyId: PropTypes.string.isRequired,
          seller: PropTypes.string.isRequired,
        })
      ),
    })
  ),
  /** Component children that will be displayed inside of the button **/
  children: PropTypes.node.isRequired,
  /** Should redirect to checkout after adding to cart */
  isOneClickBuy: PropTypes.bool,
  /** Set style to large */
  large: PropTypes.bool,
  /** Internationalization */
  intl: intlShape.isRequired,
  /** If the product is available or not*/
  available: PropTypes.bool.isRequired,
  /** Function used to show toasts (messages) to user */
  showToast: PropTypes.func.isRequired,
  /** Function to be called on the start of add to cart click event */
  onAddStart: PropTypes.func,
  /** Function to be called on the end of add to cart event */
  onAddFinish: PropTypes.func,
  /** Add to cart mutation */
  addToCart: PropTypes.func.isRequired,
}

const withMutation = graphql(
  gql`
    mutation addToCart($items: [MinicartItem]) {
      addToCart(items: $items) @client
    }
  `,
  {
    props: ({ mutate }) => ({
      addToCart: items => mutate({ variables: { items } }),
    }),
  }
)

export default compose(
  withMutation,
  withToast,
  injectIntl
)(BuyButton)
