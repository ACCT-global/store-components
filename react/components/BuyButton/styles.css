.container button > span {
  padding-left: 0;
  padding-right: 0;
}