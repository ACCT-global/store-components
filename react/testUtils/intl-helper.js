import React from 'react'
import * as reactTestingLibrary from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import defaultStrings from '../../messages/en-US.json'
const Context = React.createContext({})

export const render = (node, options) => {
  const rendered = reactTestingLibrary.render(
    <IntlProvider messages={defaultStrings} locale="en-US">
      {node}
    </IntlProvider>,
    options
  )

  return {
    ...rendered,
    rerender: newUi =>
      render(newUi, {
        container: rendered.container,
        baseElement: rendered.baseElement,
      }),
  }
}

// re-export everything
export default reactTestingLibrary
