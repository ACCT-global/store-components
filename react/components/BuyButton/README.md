# BuyButton
BuyButton is a canonical component that any VTEX app can import.

To import it into your code: 
```js
import BuyButton from 'vtex.store-components/BuyButton'
```

## Usage
You can use it in your code like a React component with the jsx tag: `<BuyButton />`. 
```jsx
<BuyButton skuItems={[{ skuId: '1', quantity: 1, seller: 1 }]}> 
  Buy!
</BuyButton>
```

| Prop name          | Type                    | Description                                                                 |
| ------------------ | ----------------------- | --------------------------------------------------------------------------- |
| `skuItems`         | `Array<SkuItem>!`       | SKU Items to be added to the cart |
| `isOneClickBuy`    | `Boolean`               | Should redirect to the checkout page or not |
| `large`            | `Boolean`               | Sets button to large style, filling whole width (like a `block`) |
| `available`        | `Boolean`               | If component is available or not |
| `showToast`        | `Boolean`               | If toast with feedback should be shown after add item request is processed |
| `onAddStart`       | `Function: () => void`  | Function called before add item request starts |
| `onAddFinish`      | `Function: () => void`  | Function called after add item request ends |

#### SKU Items Props

| Prop name          | Type                 | Description                                                                 |
| ------------------ | -------------------- | --------------------------------------------------------------------------- |
| `skuId`            | `String!`            | Specification of which product will be added to the cart |
| `quantity`         | `Number!`            | Quantity of the product sku to be added to the cart |
| `seller`           | `Number!`            | Which seller is being referenced by the button  |
| `name`             | `String!`            | Product name |
| `price`            | `Number!`            | Product price |
| `brand`            | `String!`            | Product brand |
| `variant`          | `String`             | Product variant |
| `options`          | `Array<OptionType>`  | Items to be added as assembly options of this parent product |

#### OptionType
| Prop name          | Type       | Description                                                                 |
| ------------------ | ---------- | --------------------------------------------------------------------------- |
| `id`               | `String!`  | Id of assembly option |
| `quantity`         | `Number!`  | Quantity of assembly option to be added |
| `assemblyId`       | `String!`  | parentAssemblyBinding of assembly option |
| `seller`           | `String!`  | seller of assembly option |
