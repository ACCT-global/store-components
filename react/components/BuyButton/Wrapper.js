import * as React from 'react'
import { ProductContext } from 'vtex.product-context'
import { path, isEmpty, compose } from 'ramda'
import { FormattedMessage, injectIntl } from 'react-intl'
import { withToast } from 'vtex.styleguide'
import { orderFormConsumer } from 'vtex.store-resources/OrderFormContext'

import gql from 'graphql-tag'
import { graphql } from 'react-apollo'

import { BuyButton } from './index'

const BuyButtonWrapper = ({ 
  intl,
  addToCart,
  showToast,
  orderFormContext,
  onAddStart,
  onAddFinish,
  ...props 
}) => {
  const valuesFromContext = React.useContext(ProductContext)

  const buyButtonProps = () => {
    if (!valuesFromContext || isEmpty(valuesFromContext)) return props

    const { product, selectedItem, selectedQuantity } = valuesFromContext

    const sellerId = path(['sellers', 0, 'sellerId'], selectedItem)
    const commertialOffer = path(
      ['sellers', 0, 'commertialOffer'],
      selectedItem
    )

    const showBuyButton =
      Number.isNaN(+path(['AvailableQuantity'], commertialOffer)) ||
      path(['AvailableQuantity'], commertialOffer) > 0

    return {
      skuItems: selectedItem &&
        sellerId && [
          {
            skuId: selectedItem.itemId,
            quantity: selectedQuantity,
            seller: sellerId,
            name: selectedItem.nameComplete,
            price: commertialOffer.Price,
            variant: selectedItem.name,
            brand: product.brand,
            index: 0,
            detailUrl: `/${product.linkText}/p`,
            imageUrl: path(['images', '0', 'imageUrl'], selectedItem),
            listPrice: path(
              ['sellers', '0', 'commertialOffer', 'ListPrice'],
              selectedItem
            ),
          },
        ],
      large: true,
      available: showBuyButton,
    }
  }

  const { available, ...restProps } = buyButtonProps()

  if (!available) return null

  return (
    <BuyButton
      intl={intl}
      addToCart={addToCart}
      onAddStart={onAddStart}
      onAddFinish={onAddFinish}
      showToast={showToast}
      orderFormContext={orderFormContext}
      { ...restProps }
    >
      <FormattedMessage id="store/buy-button.add-to-cart" />
    </BuyButton>
  )
}

export const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($items: [MinicartItem]) {
    addToCart(items: $items) @client
  }
`

export const OPEN_CART_MUTATION = gql`
  mutation setMinicartOpen($isOpen: Boolean!) {
    setMinicartOpen(isOpen: $isOpen) @client
  }
`

const withAddToCart = graphql(ADD_TO_CART_MUTATION, {
  name: 'addToCart',
  props: ({ addToCart }) => ({
    addToCart: items => addToCart({ variables: { items } }),
  }),
})

const withOpenMinicart = graphql(OPEN_CART_MUTATION, {
  name: 'setMinicartOpen',
  props: ({ setMinicartOpen }) => ({
    setMinicartOpen: isOpen => setMinicartOpen({ variables: { isOpen } }),
  }),
})

const withMutation = graphql(ADD_TO_CART_MUTATION, {
  props: ({ mutate }) => ({
    addToCart: items => mutate({ variables: { items } }),
  }),
})

export default compose(
  withMutation,
  withAddToCart,
  withOpenMinicart,
  withToast,
  injectIntl,
  orderFormConsumer
)(BuyButtonWrapper)
