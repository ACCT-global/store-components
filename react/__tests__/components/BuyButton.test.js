import React from 'react'
import { render } from 'intl-helper'

import BuyButton from '../../BuyButton'

describe('<BuyButton />', () => {
  const renderComponent = customProps => {
    const props = {
      ...customProps,
    }

    const comp = <BuyButton {...props}> Test </BuyButton>

    return render(comp)
  }

  it('should be rendered', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toBeDefined()
    expect(asFragment()).toBeTruthy()
  })

  it('should match snapshot normal', () => {
    const { asFragment } = renderComponent({ available: true, skuItems: [] })
    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot unavailable', () => {
    const { asFragment } = renderComponent({ available: false, skuItems: [] })
    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot loading', () => {
    const { asFragment } = renderComponent({ available: false })
    expect(asFragment()).toMatchSnapshot()
  })
})
