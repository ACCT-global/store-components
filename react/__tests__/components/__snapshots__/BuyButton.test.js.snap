// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BuyButton /> should match snapshot loading 1`] = `
<DocumentFragment>
  <svg
    class="content-loader-mock"
  />
</DocumentFragment>
`;

exports[`<BuyButton /> should match snapshot normal 1`] = `
<DocumentFragment>
  <div
    class="buyButtonContainer--available"
  >
    <button
      data-isloading="false"
    >
      Test
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BuyButton /> should match snapshot unavailable 1`] = `
<DocumentFragment>
  <div
    class="buyButtonContainer--unavailable"
  >
    <button
      data-isloading="false"
      disabled=""
    >
      <div
        class="buyButtonText"
      >
        <span>
          Unavailable
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
